cstar/roms/simulation.py

```diff
@@ -439,57 +514,165 @@ def codebases(self) -> list[ExternalCodeBase]:
         return [self.codebase, self.marbl_codebase]
 
     @property
-    def in_file(self) -> Path:
-        """Retrieves the ROMS runtime input file (.in) associated with this simulation.
+    def _forcing_paths(self) -> list[Path]:
+        """Collect and return all local paths to ROMS forcing input datasets.
+
+        This internal property gathers the `working_path` attributes of all forcing-related
+        datasets attached to the simulation — including tidal, river, surface, boundary, and
+        correction datasets — and returns a flat list of resolved file paths.
+
+        Returns
+        -------
+        list of pathlib.Path
+            A list of paths to all local forcing input files required by ROMS.
+
+        Raises
+        ------
+        ValueError
+            If any required forcing dataset does not have a `working_path` set,
+            indicating it has not been retrieved or prepared.
+
+        Notes
+        -----
+        - If any `working_path` is a list of files (e.g., multiple NetCDFs), each path
+          is included individually.
+        - This property is used to populate `runtime_settings.forcing`.
+        """
+
+        forcing_sources = chain(
+            [self.tidal_forcing, self.river_forcing],
+            self.surface_forcing,
+            self.boundary_forcing,
+            self.forcing_corrections,
+        )
+
+        forcing_paths: list[Path] = []
+
+        for source in filter(None, forcing_sources):
+            paths = source.working_path
+            if paths is None:
```

Comment:
check this outside the loop to simplify the loop itself.

```python
if [source.working_path for source in forcing_sources]:
    raise ValueError(" does not have wp")

for source in forcing_sources:
   ...
```

note: if the filter was implemented `filter(lambda x: x is not None, forcing_sources)` you could also avoid this check every time through the loop